plugins/filter_aws/aws.c

```diff
@@ -146,24 +154,54 @@ static int cb_aws_init(struct flb_filter_instance *f_ins,
         }
     }
 
-    /* v1 or v2 instance metadata */
-    ctx->use_v2 = use_v2;
+    struct flb_aws_client_generator *generator;
+    if (options && options->client_generator) {
+        generator = options->client_generator;
+        flb_plg_info(ctx->ins, "options: %s", options->name);
+    } else {
+        generator = flb_aws_client_generator();
+    }
+    ctx->client_aws = generator->create();
+    ctx->client_aws->name = "ec2_imds_provider_client";
+    ctx->client_aws->has_auth = FLB_FALSE;
+    ctx->client_aws->provider = NULL;
+    ctx->client_aws->region = NULL;
+    ctx->client_aws->service = NULL;
+    ctx->client_aws->port = FLB_AWS_IMDS_PORT;
+    ctx->client_aws->flags = 0;
+    ctx->client_aws->proxy = NULL;
+
+    struct flb_upstream *upstream;
+    upstream = flb_upstream_create(config, FLB_AWS_IMDS_HOST, FLB_AWS_IMDS_PORT,
+                                   FLB_IO_TCP, NULL);
+    if (!upstream) {
+        flb_plg_debug(ctx->ins, "unable to connect to EC2 IMDS");
+        return -1;
+    }
+
+    /* IMDSv2 token request will timeout if hops = 1 and running within container */
+    upstream->base.net.connect_timeout = FLB_AWS_IMDS_TIMEOUT;
+    upstream->base.net.io_timeout = FLB_AWS_IMDS_TIMEOUT;
+    upstream->base.net.keepalive = FLB_FALSE; /* On timeout, the connection is broken */
+    ctx->client_aws->upstream = upstream;
 
-    ctx->metadata_retrieved = FLB_FALSE;
+    ctx->client_imds = flb_aws_imds_create(&flb_aws_imds_config_default, ctx->client_aws);
+    if (!ctx->client_imds) {
+        flb_plg_error(ctx->ins, "failed to create aws client");
+        flb_free(ctx);
+        return -1;
+    }
+    ctx->client_imds->imds_version = imds_version;
 
-    ctx->ec2_upstream = flb_upstream_create(config,
-                                            FLB_FILTER_AWS_IMDS_HOST,
-                                            80,
-                                            FLB_IO_TCP,
-                                            NULL);
-    if (!ctx->ec2_upstream) {
-        flb_plg_error(ctx->ins, "connection initialization error");
+    /* Populate context with config map defaults and incoming properties */
+    ret = flb_filter_config_map_set(f_ins, (void *) ctx);
+    if (ret == -1) {
+        flb_plg_error(f_ins, "configuration error");
         flb_free(ctx);
         return -1;
     }
 
-    /* Remove async flag from upstream */
-    flb_stream_disable_async_mode(&ctx->ec2_upstream->base);
```

Comment:
I think we still need to disable async since filters can not support async?

Reply:
I don't think we do, https://github.com/fluent/fluent-bit/pull/6065/commits/b553b0a43554dcf3f012c240f435865717859fcc